File: src/main/java/com/binance/crypto/bot/api/common/response/MessageResourceResponse.java
package com.binance.crypto.bot.api.common.response;

import lombok.Getter;
import lombok.Setter;
import org.apache.commons.lang3.Validate;

import java.util.Collections;
import java.util.Map;

@Getter
@Setter
public class MessageResourceResponse<T> {

    private boolean success;
    private String messageResource;
    private String errorMessage;
    private Map<String, String> messageResourceParameters;
    private T body;

    public MessageResourceResponse() {
    }

    protected MessageResourceResponse(final boolean success, final String messageResource, final Map<String, String> messageResourceParameters, final String errorMessage) {
        this.success = success;
        this.messageResource = messageResource;
        this.messageResourceParameters = messageResourceParameters;
        this.errorMessage = errorMessage;
    }

    protected MessageResourceResponse(final boolean success, final String messageResource, final Map<String, String> messageResourceParameters, T body) {
        this.success = success;
        this.messageResource = messageResource;
        this.messageResourceParameters = messageResourceParameters;
        this.body = body;
    }

    public static <T> MessageResourceResponse<T> success(final T body) {
        return new MessageResourceResponse<>(true, null, Collections.emptyMap(), body);
    }

    public static MessageResourceResponse success() {
        return new MessageResourceResponse<>(true, null, Collections.emptyMap(), null);
    }

    public static <T> MessageResourceResponse<T> failure(final String messageResource, final Map<String, String> messageResourceParameters, final String errorMessage) {
        Validate.notNull(messageResource, "messageResource is undefined");
        return new MessageResourceResponse<>(false, messageResource, messageResourceParameters, errorMessage);
    }

    public static <T> MessageResourceResponse<T> failure(final MessageResourceException messageResourceException) {
        Validate.notNull(messageResourceException, "messageResourceException is undefined");
        Validate.notBlank(messageResourceException.getMessageResourceId(), "messageResourceException.messageResourceId is blank");
        return new MessageResourceResponse<>(false, messageResourceException.getMessageResourceId(), messageResourceException.getMessageResourceParameters(), messageResourceException.getErrorMessage());
    }

}

File: angular-ui/src/app/shared/message-resource-response.ts
export class MessageResourceResponse<DATA_TYPE> {

  private _success: boolean;
  private _messageResource: string;
  private _errorMessage: string;
  private _messageResourceParameters: Map<string, string>;
  private _body: DATA_TYPE;


  constructor(success: boolean, messageResource: string, errorMessage: string, messageResourceParameters: Map<string, string>, body: DATA_TYPE) {
    this._success = success;
    this._messageResource = messageResource;
    this._errorMessage = errorMessage;
    this._messageResourceParameters = messageResourceParameters;
    this._body = body;
  }


  get success(): boolean {
    return this._success;
  }

  set success(value: boolean) {
    this._success = value;
  }

  get messageResource(): string {
    return this._messageResource;
  }

  set messageResource(value: string) {
    this._messageResource = value;
  }

  get errorMessage(): string {
    return this._errorMessage;
  }

  set errorMessage(value: string) {
    this._errorMessage = value;
  }

  get messageResourceParameters(): Map<string, string> {
    return this._messageResourceParameters;
  }

  set messageResourceParameters(value: Map<string, string>) {
    this._messageResourceParameters = value;
  }

  get body(): DATA_TYPE {
    return this._body;
  }

  set body(value: DATA_TYPE) {
    this._body = value;
  }
}
